[develop]: Adding in the tutorial for the Halloween Storm (#1124)
ufs-community/ufs-srweather-app: doc/UsersGuide/BuildingRunningTesting/Tutorial.rst

Load the Workflow
--------------------

> To load the workflow environment, source the lmod-setup file. Then load the workflow conda environment by running:

Reviewer: ```suggestion
To load the workflow environment, source the lmod-setup file and load the workflow conda environment by running:
```

.. _CronNote:

.. include:: ../../doc-snippets/cron-note.rst

> ``EXPT_SUBDIR:`` This variable can be changed to any name the user wants from "gfsv16_physics_fcst" to "forecast1" to "a;skdfj". However, the best names will indicate useful information about the experiment. This tutorial uses ``control`` to establish a baseline, or "control", forecast. Since this tutorial helps users to compare the output from two different forecasts --- one that uses the FV3_GFS_v16 physics suite and one that uses the FV3_RRFS_v1beta physics suite --- "gfsv16_physics_fcst" could be a good alternative directory name.

Reviewer: ```suggestion
``EXPT_SUBDIR:`` This variable can be changed to any name the user wants from "gfsv16_physics_fcst" to "forecast1" to "askdfj" (but note that whitespace and some punctuation characters are not allowed). However, the best names will indicate useful information about the experiment. This tutorial uses ``control`` to establish a baseline, or "control", forecast. Since this tutorial helps users to compare the output from two different forecasts --- one that uses the FV3_GFS_v16 physics suite and one that uses the FV3_RRFS_v1beta physics suite --- "gfsv16_physics_fcst" could be a good alternative directory name.
```

**Weather Phenomena:** Flooding and high winds

   * `Storm Prediction Center (SPC) Storm Report for 20191031 <https://www.spc.noaa.gov/climo/reports/191031_rpts.html>`_

.. figure:: https://github.com/ufs-community/ufs-srweather-app/wiki/Tutorial/HalloweenStorm.gif
   :alt: Radar animation of the Halloween Storm that swept across the Eastern United States in 2019. 

   *Halloween Storm 2019*

Data
-------

Data for the Halloween Storm is publicly available in S3 data buckets. The Rapid Refresh (`RAP <https://rapidrefresh.noaa.gov/>`_) data can be downloaded from the `SRW App data bucket <https://registry.opendata.aws/noaa-ufs-shortrangeweather/>`_ using ``wget``. Make sure to issue the command from the folder where you want to place the data.

.. code-block:: console

   wget https://noaa-ufs-srw-pds.s3.amazonaws.com/develop-20240618/halloween_rap.tgz
   tar -xzf halloween_rap.tgz
   
This will untar the ``halloween_rap.tgz`` data into a directory named ``RAP``.

The SRW App can pull HRRR data directly from the `HRRR data bucket <https://registry.opendata.aws/noaa-hrrr-pds/>`_. Users do not need to download the data separately. 

Load the workflow
---------------------

> To load the workflow environment, source the lmod-setup file. Then load the workflow conda environment by running:

Reviewer: ```suggestion
To load the workflow environment, source the lmod-setup file and load the workflow conda environment by running:
```

Edit the configuration file (``config.yaml``) to include the variables and values in the sample configuration excerpts below. 

.. include:: ../../doc-snippets/file-edit-hint.rst

Start in the ``user:`` section and change the ``MACHINE`` and ``ACCOUNT`` variables. For example, when running on a personal MacOS device, users might set:

.. code-block:: console

   user:
      RUN_ENVIR: community
      MACHINE: macos
      ACCOUNT: none

For a detailed description of these variables, see :numref:`Section %s <user>`.

Users do not need to change the ``platform:`` section of the configuration file for this tutorial. 

In the ``workflow:`` section of ``config.yaml``, update ``EXPT_SUBDIR``, ``CCPP_PHYS_SUITE``, ``PREDEF_GRID_NAME``, ``DATE_FIRST_CYCL``, ``DATE_LAST_CYCL``, and ``FCST_LEN_HRS``.

.. code-block:: console

   workflow:
     USE_CRON_TO_RELAUNCH: false
     EXPT_SUBDIR: halloweenRAP
     CCPP_PHYS_SUITE: FV3_RAP
     PREDEF_GRID_NAME: RRFS_CONUS_13km
     DATE_FIRST_CYCL: '2019103012'
     DATE_LAST_CYCL: '2019103012'
     FCST_LEN_HRS: 36
     PREEXISTING_DIR_METHOD: rename
     VERBOSE: true
     COMPILER: intel

.. include:: ../../doc-snippets/cron-note.rst

> ``EXPT_SUBDIR:`` This variable can be changed to any name the user wants from "halloweenRAP" to "HalloweenStorm1" to "a;skdfj". However, the best names will indicate useful information about the experiment. Since this tutorial helps users to compare the output from RAP and HRRR forecast input data, this tutorial will use ``halloweenRAP`` for the Halloween Storm experiment that uses RAP forecast data.

Reviewer: ```suggestion
``EXPT_SUBDIR:`` This variable can be changed to any name the user wants from "halloweenRAP" to "HalloweenStorm1" to "askdfj" (but note that whitespace and some punctuation characters are not allowed). However, the best names will indicate useful information about the experiment. Since this tutorial helps users to compare the output from RAP and HRRR forecast input data, this tutorial will use ``halloweenRAP`` for the Halloween Storm experiment that uses RAP forecast data.
```

In the ``task_get_extrn_ics:`` section, add ``USE_USER_STAGED_EXTRN_FILES`` and ``EXTRN_MDL_SOURCE_BASEDIR_ICS``. Users will need to adjust the file path to point to the location of the data on their system. 

.. code-block:: console

   task_get_extrn_ics:
     EXTRN_MDL_NAME_ICS: RAP
     USE_USER_STAGED_EXTRN_FILES: true
     EXTRN_MDL_SOURCE_BASEDIR_ICS: /path/to/RAP/for_ICS

For a detailed description of the ``task_get_extrn_ics:`` variables, see :numref:`Section %s <task_get_extrn_ics>`. 

Similarly, in the ``task_get_extrn_lbcs:`` section, add ``USE_USER_STAGED_EXTRN_FILES`` and ``EXTRN_MDL_SOURCE_BASEDIR_LBCS``. Users will need to adjust the file path to point to the location of the data on their system. 

.. code-block:: console

   task_get_extrn_lbcs:
     EXTRN_MDL_NAME_LBCS: RAP
     LBC_SPEC_INTVL_HRS: 3
     USE_USER_STAGED_EXTRN_FILES: true
     EXTRN_MDL_SOURCE_BASEDIR_LBCS: /path/to/RAP/for_LBCS

For a detailed description of the ``task_get_extrn_lbcs:`` variables, see :numref:`Section %s <task_get_extrn_lbcs>`. 

Users do not need to modify the ``task_run_fcst:`` section for this tutorial. 

> .. COMMENT: Do we need to set QUILTING to true?

Reviewer 2: QUILTING is required for regional grids.  The default setting is true in `ush/config_defaults.yaml`.  Further, `QUILTING: true` is being set in `config.community.yaml`, so there should be no need to modify the `task_run_fcst:` section for the tutorial.

.. code-block:: console

   task_get_extrn_ics:
     EXTRN_MDL_NAME_ICS: HRRR
     USE_USER_STAGED_EXTRN_FILES: false
     EXTRN_MDL_FILES_ICS:
     - '{yy}{jjj}{hh}00{fcst_hr:02d}00'

For a detailed description of the ``task_get_extrn_ics:`` variables, see :numref:`Section %s <task_get_extrn_ics>`. 

Update the same values in the ``task_get_extrn_lbcs:`` section: 

.. code-block:: console

   task_get_extrn_lbcs:
     EXTRN_MDL_NAME_LBCS: HRRR
     LBC_SPEC_INTVL_HRS: 3
     USE_USER_STAGED_EXTRN_FILES: false
     EXTRN_MDL_FILES_LBCS:
     - '{yy}{jjj}{hh}00{fcst_hr:02d}00'

For a detailed description of the ``task_get_extrn_lbcs:`` variables, see :numref:`Section %s <task_get_extrn_lbcs>`. 

After configuring the forecast, users can generate the second forecast by running:

.. code-block:: console

   ./generate_FV3LAM_wflow.py

To see experiment progress, users should navigate to their experiment directory. As in the first forecast, the following commands allow users to launch new workflow tasks and check on experiment progress. 

.. code-block:: console

>    cd /path/to/expt_dirs/test_expt

Reviewer: ```suggestion
   cd /path/to/expt_dirs/halloweenHRRR
```

In the 250mb wind plots below, the ``halloweenHRRR`` and ``halloweenRAP`` plots are nearly identical at forecast hour f036. This shows great model agreement. Analyzing this chart we can see multiple ingredients signaling a significant severe weather event over the eastern CONUS. The first thing to notice is the placement of the jet streak along with troughing approaching the eastern US. Also notice an extreme 150KT jet max over southern Ohio further fueling severe weather. The last thing to notice is the divergence aloft present over the eastern CONUS; seeing divergence present all the way up to 250mb indicates a strong system.

.. figure:: https://github.com/ufs-community/ufs-srweather-app/wiki/fcst4_plots/250wind_rap_conus_f036.png
      :align: center
      :width: 75%

      *RAP Plot for 250mb Wind*

.. figure:: https://github.com/ufs-community/ufs-srweather-app/wiki/fcst4_plots/250wind_hrrr_conus_f036.png
      :align: center
      :width: 75%

      *HRRR Plot for 250mb Wind*

.. _fcst4_10mwind:

10m Wind
``````````
The 10m wind plots allows forecasters to pick up on patterns closer to the surface. It shows features such as convergence and pressure areas.

> In the 10m wind plots below, the ``halloweenHRRR`` and ``halloweenRAP`` are once again very similar, which makes sense given that the 250mb wind plots are also so similar. We can see a few key feautres on this chart. The most important is the area of convergence taking place over the east coast which is driving the line of severe storms. 

Reviewer 2: features

```suggestion
In the 10m wind plots below, the ``halloweenHRRR`` and ``halloweenRAP`` are once again very similar, which makes sense given that the 250mb wind plots are also so similar. We can see a few key features on this chart. The most important is the area of convergence taking place over the east coast which is driving the line of severe storms. 
```

.. figure:: https://github.com/ufs-community/ufs-srweather-app/wiki/fcst4_plots/10mwind_rap_conus_f036.png
      :align: center
      :width: 75%

>       *Rap Plot for 10m Winds*

Reviewer 2: RAP

```suggestion
      *RAP Plot for 10m Winds*
```

**Weather Phenomena:** Flooding and high winds

>    * `Storm Prediction Center (SPC) Storm Report for 20191031 

Reviewer: It looks like the URL got deleted.
```suggestion
   * `Storm Prediction Center (SPC) Storm Report for 20191031 <https://www.spc.noaa.gov/climo/reports/191031_rpts.html>`_
```

.. figure:: https://github.com/ufs-community/ufs-srweather-app/wiki/Tutorial/HalloweenStorm.jpg
   :alt: Radar animation of the Halloween Storm that swept across the Eastern United States in 2019. 

   *Halloween Storm 2019*

Data
-------

> On :srw-wiki:`Level 1 <Supported-Platforms-and-Compilers>` systems, users can find data for the Halloween Storm in the usual input model data locations (see :numref:`Section %s <DataLocations>` for a list). The data can also be downloaded from the `UFS SRW Halloween Storm Case Study <https://ufs-case-studies.readthedocs.io/en/develop/2019HaloweenStorm.html>`__.

Reviewer: ```suggestion
On :srw-wiki:`Level 1 <Supported-Platforms-and-Compilers>` systems, users can find data for the Halloween Storm in the usual input model data locations (see :numref:`Section %s <DataLocations>` for a list). The RAP data can also be downloaded from the `SRW App data bucket <https://registry.opendata.aws/noaa-ufs-shortrangeweather/>`_ using ``wget``:

.. code-block:: console

   wget https://noaa-ufs-srw-pds.s3.amazonaws.com/develop-20240618/halloween_rap.tgz
   tar -xzf halloween_rap.tgz
```
You'll also need to add info about getting the HRRR data.

Load the workflow
---------------------

To load the workflow environment, source the lmod-setup file. Then load the workflow conda environment. From the ``ufs-srweather-app`` directory, run:

.. code-block:: console
   

>    source etc/lmod-setup.sh <platform>       # OR: source etc/lmod-setup.csh <platform> when running in a csh/tcsh shell

Reviewer: In general, I think it's best to phase out support for csh. We are already assuming bash shell in most other places. 
```suggestion
   source etc/lmod-setup.sh <platform>
```

   module use modulefiles
   module load wflow_<platform>

where ``<platform>`` is a valid, lowercased machine name (see ``MACHINE`` in :numref:`Section %s <user>` for valid values). 

After loading the workflow, users should follow the instructions printed to the console. Usually, the instructions will tell the user to run |activate|. For example, a user on Hera with permissions on the ``nems`` project may issue the following commands to load the workflow (replacing ``User.Name`` with their actual username):

.. code-block:: console
   
   source /scratch1/NCEPDEV/nems/User.Name/ufs-srweather-app/etc/lmod-setup.sh hera
   module use /scratch1/NCEPDEV/nems/User.Name/ufs-srweather-app/modulefiles
   module load wflow_hera

>    conda activate srw_app

Reviewer: Might want to change to Gaea bc technically, the `source lmod-setup.sh` line is not required on Hera (tho it won't mess anything up). Alternatively, maybe we should delete the example, since it's already in the tutorial above, and there's a brief explanation here? Don't want to be too repetitive.

> Configurations

Reviewer: ```suggestion
Configuration
```

-------------------------

Navigate to the ``ufs-srweather-app/ush`` directory. The default (or "control") configuration for this experiment is based on the ``config.community.yaml`` file in that directory. Users can copy this file into ``config.yaml`` if they have not already done so:

.. code-block:: console

   cd /path/to/ufs-srweather-app/ush
   

Users can save the location of the ``ush`` directory in an environment variable (``$USH``). This makes it easier to navigate between directories later. For example:

.. code-block:: console

   export USH=/path/to/ufs-srweather-app/ush

> Users should substitute ``/path/to/ufs-srweather-app/ush`` with the actual path on their system. As long as a user remains logged into their system, they can run ``cd $USH``, and it will take them to the ``ush`` directory. The variable will need to be reset for each login session. 

Reviewer: This is a repeated code snippet that we might want to remove from both tutorials, stick in a separate file, and add via an `..include::` directive here.

.. code-block:: console

   user:
      RUN_ENVIR: community
      MACHINE: macos
      ACCOUNT: none

For a detailed description of these variables, see :numref:`Section %s <user>`.

Users do not need to change the ``platform:`` section of the configuration file for this tutorial. The default parameters in the ``platform:`` section pertain to METplus verification, which is not addressed here. For more information on verification, see :numref:`Section %s <VXCases>`.

In the ``workflow:`` section of ``config.yaml``, update ``EXPT_SUBDIR`` and ``PREDEF_GRID_NAME``.

.. code-block:: console

   workflow:
     USE_CRON_TO_RELAUNCH: false
     EXPT_SUBDIR: control
     CCPP_PHYS_SUITE: FV3_GFS_v16
     PREDEF_GRID_NAME: RRFS_CONUS_13km
     DATE_FIRST_CYCL: '2019102812'
     DATE_LAST_CYCL: '2019102812'
     FCST_LEN_HRS: 12
     PREEXISTING_DIR_METHOD: rename
     VERBOSE: true
     COMPILER: intel

.. _CronNote:

.. note::

   Users may also want to set ``USE_CRON_TO_RELAUNCH: true`` and add ``CRON_RELAUNCH_INTVL_MNTS: 3``. This will automate submission of workflow tasks when running the experiment. However, not all systems have :term:`cron`. 

> ``EXPT_SUBDIR:`` This variable can be changed to any name the user wants from "gfsv16_physics_fcst" to "forecast1" to "a;skdfj". However, the best names will indicate useful information about the experiment. This tutorial uses ``control`` to establish a baseline, or "control", forecast. Since this tutorial helps users to compare the output from two different forecasts --- one that uses the FV3_GFS_v16 physics suite and one that uses the FV3_RRFS_v1beta physics suite --- "gfsv16_physics_fcst" could be a good alternative directory name.

Reviewer: Are we comparing output from different physics suites? I thought the physics was the same for both. I would adapt this description to be more appropriate for this particular experiment, e.g., use `rap_fcst` and `hrrr_fcst`.

> ``PREDEF_GRID_NAME:`` This experiment uses the RRFS_CONUS_13km, rather than the default RRFS_CONUS_25km grid. Using the RRFS_CONUS_13km grid allows for fewer computational restraints compared to the 25km grid.  For more information on this grid, see :numref:`Section %s <RRFS_CONUS_13km>`.

Reviewer: ```suggestion
``PREDEF_GRID_NAME:`` This experiment uses the RRFS_CONUS_13km, rather than the default RRFS_CONUS_25km grid. Using the RRFS_CONUS_13km grid provides a higher resolution forecast, more detailed forecast; however, it is more computationally expensive compared to the 25km grid.  For more information on this grid, see :numref:`Section %s <RRFS_CONUS_13km>`.
```
@jdkublnick Technically, there are more computational restraints with `RRFS_CONUS_13km` because it requires more compute power to run at a higher resolution (note that higher resolution translates to smaller grid cells and more of them. 13km is the approximate grid cell size, and there will be more of those on the CONUS compared to the 25km grid cells on the lower resolution `RRFS_CONUS_25km` domain). 
@MichaelLueken Does my change above make sense, or can you refine it?

Reviewer 2: @gspetro-NOAA - Your modification makes sense.  The use of the 13km grid versus the 25km grid is for a higher resolution, more detailed forecast.  The 13 km grid will also require more resources in order to run, for the reason you noted above.  The only modification I would make would be to remove forecast after higher resolution:
```suggestion
``PREDEF_GRID_NAME:`` This experiment uses the RRFS_CONUS_13km, rather than the default RRFS_CONUS_25km grid. Using the RRFS_CONUS_13km grid provides a higher resolution, more detailed forecast; however, it is more computationally expensive compared to the 25km grid.  For more information on this grid, see :numref:`Section %s <RRFS_CONUS_13km>`.
```

For a detailed description of other ``workflow:`` variables, see :numref:`Section %s <workflow>`.

To turn on the plotting for the experiment, the YAML configuration file
should be included in the ``rocoto:tasks:taskgroups:`` section, like this:

.. code-block:: console

 rocoto:
    tasks:
      task_get_extrn_ics:
        walltime: 06:00:00
      task_get_extrn_lbcs:
        walltime: 06:00:00
      metatask_run_ensemble:
        task_make_lbcs_mem#mem#:
          walltime: 06:00:00
        task_run_fcst_mem#mem#:
          walltime: 06:00:00

>         taskgroups: '{{ ["parm/wflow/prep.yaml", "parm/wflow/coldstart.yaml", "parm/wflow/post.yaml", "parm/wflow/plot.yaml"]|include }}'

Reviewer: Not sure if this indent is correct? It should probably line up with `metatask_run_ensemble`.

For more information on how to turn on/off tasks in the workflow, please
see :numref:`Section %s <ConfigTasks>`.

In the ``task_get_extrn_ics:`` section, add ``USE_USER_STAGED_EXTRN_FILES`` and ``EXTRN_MDL_SOURCE_BASEDIR_ICS``. Users will need to adjust the file path to reflect the location of data on their system (see :numref:`Section %s <Data>` for locations on :srw-wiki:`Level 1 <Supported-Platforms-and-Compilers>` systems). 

.. code-block:: console

   task_get_extrn_ics:
     EXTRN_MDL_NAME_ICS: UFS-CASE-STUDY
     FV3GFS_FILE_FMT_ICS: nemsio
     USE_USER_STAGED_EXTRN_FILES: false

For a detailed description of the ``task_get_extrn_ics:`` variables, see :numref:`Section %s <task_get_extrn_ics>`. 

Similarly, in the ``task_get_extrn_lbcs:`` section, add ``USE_USER_STAGED_EXTRN_FILES`` and ``EXTRN_MDL_SOURCE_BASEDIR_LBCS``. Users will need to adjust the file path to reflect the location of data on their system (see :numref:`Section %s <Data>` for locations on Level 1 systems). 

.. code-block:: console

   task_get_extrn_lbcs:
     EXTRN_MDL_NAME_LBCS: UFS-CASE-STUDY
     LBC_SPEC_INTVL_HRS: 3
     FV3GFS_FILE_FMT_LBCS: nemsio

>      USE_USER_STAGED_EXTRN_FILES: false

Reviewer: This needs to be updated to reflect the RAP data, not GFS.
